# This script contains the following chapters:
1. Importing libraries and data
2. Data Cleaning & Basic descriptive statistics

1. Importing libraries and data

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating path for easier import
path = r'C:\Users\magia\NY CitiBike'

In [3]:
# importing dataset
df = pd.read_csv(os.path.join(path, '2. Data', 'Original Data', 'citibike.csv'))

In [4]:
#inspecting data with shape attribute
df.shape

(50000, 18)

In [5]:
# inspecting data with head method
df.head()

trip_id  bike_id weekday  start_hour           start_time  start_station_id  \
0  LnQzQk    16013     Mon          18  2013-09-09 18:18:55               523   
1  IL9boN    15230     Thu          18  2013-09-12 18:38:53               257   
2  46clGB    17942     Wed          19  2013-09-18 19:44:04               479   
3  v7vdFt    19683     Sat          11  2013-09-28 11:54:37               527   
4  VGBsb5    18024     Sat          18  2013-09-07 18:08:22               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

              end_time  end_station_id     end_station_name  \
0  2013-09-09 18:35:28             334      W 20 St & 7 Ave   
1  2013-09-12 18:48:34             236  St Marks Pl & 2 Ave   
2  2013-09-18 19:50:05             513     W 56 St & 10 Ave   
3  2013-09-28 12:03:58             441      E 52 St & 2 Ave   
4  2013-09-07 18:46:38             476      E 31 St & 3 Ave   

   end_station_latitude  end_station_longitude  trip_duration      subscriber  \
0             40.742388             -73.997262            993      Subscriber   
1             40.728419             -73.987140            581      Subscriber   
2             40.768254             -73.988639            361      Subscriber   
3             40.756014             -73.967416            561      Subscriber   
4             40.743943             -73.979661           2296  Non-Subscriber   

   birth_year  gender  
0      1968.0       2  
1      1983.0       1  
2      1989.0       1  
3      1988.0       2  
4         NaN       0

2. Data Cleaning & Basic descriptive statistics

In [6]:
# checking for duplicates
df.duplicated().any()

False

In [7]:
# dropping trip_id column
df.drop('trip_id', axis=1, inplace=True)

In [8]:
# inspecting data with info method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bike_id                  50000 non-null  int64  
 1   weekday                  50000 non-null  object 
 2   start_hour               50000 non-null  int64  
 3   start_time               50000 non-null  object 
 4   start_station_id         50000 non-null  int64  
 5   start_station_name       50000 non-null  object 
 6   start_station_latitude   50000 non-null  float64
 7   start_station_longitude  50000 non-null  float64
 8   end_time                 50000 non-null  object 
 9   end_station_id           50000 non-null  int64  
 10  end_station_name         50000 non-null  object 
 11  end_station_latitude     50000 non-null  float64
 12  end_station_longitude    50000 non-null  float64
 13  trip_duration            50000 non-null  int64  
 14  subscriber            

In [9]:
# checking values of birth_year column
df['birth_year'].unique()

array([1968., 1983., 1989., 1988.,   nan, 1986., 1994., 1969., 1993.,
       1985., 1979., 1967., 1981., 1959., 1976., 1974., 1978., 1971.,
       1966., 1987., 1965., 1951., 1954., 1960., 1984., 1955., 1958.,
       1975., 1953., 1980., 1973., 1957., 1970., 1977., 1961., 1982.,
       1962., 1950., 1956., 1991., 1964., 1946., 1947., 1963., 1952.,
       1992., 1990., 1972., 1944., 1922., 1997., 1941., 1995., 1932.,
       1900., 1948., 1945., 1949., 1901., 1996., 1938., 1942., 1943.,
       1935., 1940., 1939., 1933., 1899., 1934., 1936., 1910., 1929.,
       1917., 1937., 1926., 1924., 1921.])

In [10]:
#checking %  of NaN in birth_year
df['birth_year'].isna().mean() * 100

13.958

In [11]:
# creating birth_year_label column to show NaNs as "Unknown" and numeric birth years as integers
df['birth_year_label'] = df['birth_year'].fillna('Unknown')

# Convert numeric years to int, leave 'Unknown' as is
def clean_year(x):
    if x == 'Unknown':
        return x
    else:
        return str(int(x))

df['birth_year_label'] = df['birth_year_label'].apply(clean_year)

In [12]:
# convert gender to readable labels
gender_map = {0: 'Unknown', 1: 'Male', 2: 'Female'}
df['gender'] = df['gender'].map(gender_map)

print(df['gender'].value_counts())


gender
Male       32361
Female     10658
Unknown     6981
Name: count, dtype: int64


In [13]:
# converting start_time and end_time data from string object to datetime
df['start_time'] = pd.to_datetime(df['start_time'])
df['end_time'] = pd.to_datetime(df['end_time'])

In [14]:
# checking weekday column
print(df['weekday'].value_counts())

weekday
Mon    7715
Sun    7319
Fri    7309
Wed    7228
Tue    7065
Thu    6857
Sat    6507
Name: count, dtype: int64


In [15]:
# checking subscriber column
print(df['subscriber'].value_counts())

subscriber
Subscriber        43021
Non-Subscriber     6979
Name: count, dtype: int64


In [16]:
# checking for anomalies in numeric data
print("Trip duration <= 0:", df[df['trip_duration'] <= 0].shape[0])
print("Start hour outside 0-23:", df[(df['start_hour'] < 0) | (df['start_hour'] > 23)].shape[0])
print("Birth year outside valid range:", df[(df['birth_year'] > 2013) | (df['birth_year'] < 1925)].shape[0])

Trip duration <= 0: 0
Start hour outside 0-23: 0
Birth year outside valid range: 29


In [17]:
#creating  88+ years old flag for particularly senior users(potential mistakes in data)
df['88+_years_old'] = df['birth_year'] < 1925

In [18]:
# checking latitute and longitude for start station columns
invalid_lat = df[(df['start_station_latitude'] < 40.5) | (df['start_station_latitude'] > 41.0)]
invalid_lon = df[(df['start_station_longitude'] < -74.3) | (df['start_station_longitude'] > -73.7)]

print(f"Invalid latitudes: {len(invalid_lat)}")
print(f"Invalid longitudes: {len(invalid_lon)}")

Invalid latitudes: 0
Invalid longitudes: 0


In [19]:
# checking latitude and longitude for end station columns

invalid_end_lat = df[(df['end_station_latitude'] < 40.5) | (df['end_station_latitude'] > 41.0)]
invalid_end_lon = df[(df['end_station_longitude'] < -74.3) | (df['end_station_longitude'] > -73.7)]

print(f"Invalid end station latitudes: {len(invalid_end_lat)}")
print(f"Invalid end station longitudes: {len(invalid_end_lon)}")

Invalid end station latitudes: 0
Invalid end station longitudes: 0


In [20]:
# descriptive statistics for start hour, trip duration and birth year
df[['start_hour', 'trip_duration', 'birth_year']].describe()

start_hour  trip_duration    birth_year
count  50000.000000   50000.000000  43021.000000
mean      14.145240     838.982900   1975.627786
std        4.860541     573.663997     11.089001
min        0.000000      60.000000   1899.000000
25%       10.000000     417.000000   1968.000000
50%       15.000000     672.000000   1978.000000
75%       18.000000    1112.000000   1984.000000
max       23.000000    2697.000000   1997.000000

- Standard deviation of start_hour of about 4.9 hours indicates variability throughout the day.25% of trips start before 10 AM, 50% before 3 PM, and 75% before 6 PM. The busiest time of day is thus between 3 PM and 6 PM (3 hours getting 25% of the rides)
- Mean trip_duration is about 840 seconds. Large standard deviation (573.66), showing trip lengths vary widely.Maximum trip duration is 2697 seconds (almost 45 minutes).
- birth_year:Only 43,021 non-missing records out of 50,000 (since  6,979 records are NaNs).
Mean birth year is 1975.63. Standard deviation around 11 years, indicating a fairly broad age range. Birth years range from 1899 (!) to 1997.
Quartiles: 25% born before 1968, median 1978, 75% born before 1984.

In [21]:
# exporting data
df.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'citibike_cleaned.csv'), index=False)